### PR TITLE
Add controller retry support scaffolding

### PR DESCRIPTION
**Pull request by ironcladlou**
Introduce a reusable controller package which facilitates safe retries
for controllers which make use of cache.FIFO.

Port the DeploymentConfigController to use the new support as a reference
implementation of the pattern.

### CODE REVIEW

**Comment by ironcladlou** [comment]
@pmorie @bparees PTAL.

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+// exceeded, at which point resource will be forgotten and no longer retried.
+func (r *QueueRetryManager) Retry(resource interface{}) {
+	id, _ := r.keyFunc(resource)
+
+	if _, exists := r.retries[id]; !exists {
+		r.retries[id] = 0
+	}
+	tries := r.retries[id]
+
+	if tries < r.maxRetries {
```
Comment (bparees) [inline pkg/controller/controller.go]: let's add support for maxRetries==-1 meaning "forever"

Comment (ironcladlou) [inline pkg/controller/controller.go]: Done. Also, I hate using a constructor func rather than field initialization for this thing, but then the map would have to be lazily initialized. Ugh!

**Comment by bparees** [comment]
Needs more testing around the retry behavior, but the implementation looks sound to me.  I am slightly biased though.

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	go kutil.Until(func() { c.RunOnce() }, 0, c.Stop)
+}
+
+// RunOnce calls Handle. If Handle returns a retryable error, the handled
+// resource is passed to the RetryManager. If no error is returned from
+// Handle, the RetryManager is asked to forget the processed resource.
+func (c *RetryController) RunOnce() {
+	resource, err := c.HandleNext()
+	if err != nil {
+		if _, isRetryable := err.(*RetryableError); isRetryable {
```
Comment (smarterclayton) [inline pkg/controller/controller.go]: Don't do this.

Comment (smarterclayton) [inline pkg/controller/controller.go]: Make the decision to retry be an interface passed to RetryController

Comment (ironcladlou) [inline pkg/controller/controller.go]: Good idea... done. Was on the fence about actually defining an interface, ended up just passing a function and providing a stock implementation.

**Comment by ironcladlou** [comment]
Note: Before a merge, I want to add more tests for this (especially integration tests). To avoid too much rework on my part, I'm putting them off for now until we have more consensus.

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+// MaxRetries number of times.
+type QueueRetryManager struct {
+	// queue is where resources are re-queued.
+	queue Queue
+
+	// keyFunc is used to index resources.
+	keyFunc kcache.KeyFunc
+
+	// maxRetries is the total number of attempts to requeue an individual
+	// resource before giving up.
```
Comment (bparees) [inline pkg/controller/controller.go]: // a maxRetries value of -1 is interpreted as retry forever.

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	maxRetries int
+
+	// retries maps resources to their total retry count.
```
Comment (bparees) [inline pkg/controller/controller.go]: s/total/current/   (total sounds a little too close to max)

**Comment by bparees** [comment]
absurdly minor nits on my part, lgtm.

**Comment by ironcladlou** [comment]
I added a nice test, `TestRetryController_realFifoEventOrdering`, which does a miniature integration-like test with the RetryController+QueueRetryManager+FIFO to demonstrate event order guarantees with an  asynchronous queue producer. Unless there are any other special requests, I feel like this is in good shape for another merge review. Going to run it through a jenkins [test].

**Comment by pmorie** [comment]
Provisionally LGTM but I need to review the latest code.

File: Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache/fifo_test.go [hunk 5]
```diff
@@ -160,3 +160,27 @@ func TestFIFO_detectLineJumpers(t *testing.T) {
 		t.Fatalf("expected %d, got %d", e, a)
 	}
 }
+
```
Comment (pmorie) [inline Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client/cache/fifo_test.go]: Should there be a test like TestFIFO_basic that randomly does an `AddIfNotPresent` instead of just `Add` ?

**Comment by pmorie** [comment]
I really like the retry controller thing, this is informing my work on delta-based stores and controllers that do this.

**Comment by soltysh** [comment]
@ironcladlou great work with this :+1:

File: pkg/controller/controller.go [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+	c.RetryManager.Forget(resource)
+}
+
+// RetryableError is an error that is recoverable and which should
+// signal another attempt at processing a resource.
+type RetryableError struct {
+	err error
+}
+
+func (e *RetryableError) Error() string {
```
Comment (smarterclayton) [inline pkg/controller/controller.go]: Don't need these anymore

Comment (ironcladlou) [inline pkg/controller/controller.go]: Can you elaborate a little more? The default ShouldRetry impl is `RetryOnRetryableError` which checks for `RetryableError`, and the deployment controllers return it.

Comment (smarterclayton) [inline pkg/controller/controller.go]: The deployment controllers should be implementing their own retry behavior which is either retry on error, or return a special type.  Wrapping an error like this is questionable - instead handle should return opaque errors, and the controller retry handler should make a decision and either log "not retrying" or "retrying".  Using someone else's error is coupling unrelated code.

Comment (bparees) [inline pkg/controller/controller.go]: if you force me to implement my own retry behavior in my controller, it's going to look exactly like this default implementation...  i'm going to return an error that clearly indicates "retry me" and write a retrier that recognizes such errors.  i'm not going to separate my logic such that in one place i'm trying to do things, and in another place i'm trying to figure out whether i should retry that thing.

Comment (ironcladlou) [inline pkg/controller/controller.go]: If someone using the RetryController wants to establish their own error taxonomy and retry handler around it, they're free to do so and the abstractions here will facilitate them. However, all our existing use cases in deployment and builds (I think) can use this OOTB error type and handler for now, IMO.

I could me misunderstanding the proposed alternative design... maybe try once more to clarify just to be sure?

Comment (ironcladlou) [inline pkg/controller/controller.go]: I think I might see what you mean now, @smarterclayton. Will add a new commit for discussion based on what you're saying.

**Comment by ironcladlou** [comment]
@smarterclayton @bparees @pmorie 

I added a temporary commit to demonstrate a different way of dealing with errors. The new approach addresses @smarterclayton's feedback (I think), but also incorporates a strategy I learned about from the Go standard library based on [this very relevant golang-nuts discussion](https://groups.google.com/forum/#!msg/golang-nuts/8uHzrY1rZMw/F45f_8MvhOUJ). Check out the way custom errors and error transience are handled in [the golang net package](https://golang.org/src/net/net.go). I used the stdlib approach for error definition/taxonomy, and used @smarterclayton's idea to move the decision-making to the factory's `ShouldRetry` implementation.

Let me know what you think. I won't bother changing any tests until we get this part nailed down.

**Comment by pmorie** [comment]
@ironcladlou I like it. :facepunch:

**Comment by ironcladlou** [comment]
@smarterclayton @bparees
The latest revision moves DeploymentConfigController/Factory into its own package. DeploymentConfigController has a private fatal error type which the factory recognizes to drive retries in a simple manner.

Still haven't touched tests yet. When the structure of this one controller looks good to all, I'll fix the tests, get this PR merged, and then start porting all the other deploy controllers in followup PRs.

**Comment by smarterclayton** [comment]
I'm about to take off - that was my last substantive issue (no one outside of controller and its tests should know about the errors, and caking util.HandleError before you eat one).

**Comment by ironcladlou** [comment]
[test]

**Comment by pmorie** [comment]
Going to take a last pass at this before merge

**Comment by pmorie** [comment]
LGTM

[merge]

File: pkg/deploy/controller/deploymentconfig/factory.go [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+	}
+
+	return &controller.RetryController{
+		Queue:        queue,
+		RetryManager: controller.NewQueueRetryManager(queue, cache.MetaNamespaceKeyFunc, 1),
+		ShouldRetry: func(obj interface{}, err error) bool {
+			if _, isFatal := err.(fatalError); isFatal {
+				return false
+			}
+			kutil.HandleError(err)
```
Comment (bparees) [inline pkg/deploy/controller/deploymentconfig/factory.go]: seems backwards that you're logging the error when you retry, but not logging it if it's fatal.

Comment (smarterclayton) [inline pkg/deploy/controller/deploymentconfig/factory.go]: Yeah that's backwards. Logging transients is good, but fatal is critical.

Comment (ironcladlou) [inline pkg/deploy/controller/deploymentconfig/factory.go]: I caught and fixed this in the followup (#1170). Good eye!

Comment (bparees) [inline pkg/deploy/controller/deploymentconfig/factory.go]: good but slow apparently :)
